codespan-reporting/src/term.rs

```diff
@@ -21,12 +21,56 @@ pub fn emit<Source: AsRef<str>>(
     files: &Files<Source>,
     diagnostic: &Diagnostic,
 ) -> io::Result<()> {
-    use self::views::{RichDiagnostic, ShortDiagnostic};
+    use std::collections::BTreeMap;
+
+    use self::views::MarkStyle;
+    use self::views::{Header, NewLine, Note, SourceSnippet};
+
+    // Emit the title
+    //
+    // ```text
+    // error[E0001]: unexpected type in `+` application
+    // ```
+    if let Some(title) = &diagnostic.title {
+        Header::new(diagnostic.severity, title).emit(writer, config)?;
+        NewLine::new().emit(writer, config)?;
+    }
+
+    // Group labels by file
+
+    let mut label_groups = BTreeMap::new();
 
-    match config.display_style {
-        DisplayStyle::Rich => RichDiagnostic::new(diagnostic).emit(files, writer, config),
-        DisplayStyle::Short => ShortDiagnostic::new(diagnostic).emit(files, writer, config),
+    for label in &diagnostic.labels {
+        let mark_style = match label.style {
+            LabelStyle::Primary => MarkStyle::Primary(diagnostic.severity),
+            LabelStyle::Secondary => MarkStyle::Secondary,
+        };
+        label_groups
+            .entry(label.file_id)
+            .or_insert(vec![])
```

Comment:
```suggestion
            .or_default()
```

Reply:
Tbh I've always been a tad suspicious of using `Default` for stuff like monoidal identities… kind of would prefer to leave it explicit. 😅

Reply:
The main nit is that `or_insert(vec![])` creates an (empty) `Vec` regardless of if the id exists. `or_insert_with(Vec::new)` is equivalent to `or_default` but a bit longer.

I'd agree if the value were generic, but since the value is a concret `Vec` I don't find `or_default` any surprising :man_shrugging:

Reply:
Hmm, I guess I thought that seeing as `Vec::new()` doesn't allocate, it would be fine? But I do see your point.

Reply:
Yeah, it is very unlikely to matter performance wise. It is more of a clippy warning I'd say.